libportal/notification.c

```diff
@@ -131,33 +222,43 @@ xdp_portal_add_notification (XdpPortal *portal,
 {
   GAsyncReadyCallback call_done_cb = NULL;
   CallDoneData *call_done_data = NULL;
+  GUnixFDList *fd_list = NULL;
+  GVariantBuilder builder;
 
   g_return_if_fail (XDP_IS_PORTAL (portal));
   g_return_if_fail (flags == XDP_NOTIFICATION_FLAG_NONE);
 
   ensure_action_invoked_connection (portal);
 
+  g_variant_builder_init (&builder, G_VARIANT_TYPE ("(sa{sv})"));
+  g_variant_builder_add (&builder, "s", id);
+
+  g_variant_builder_open (&builder, G_VARIANT_TYPE ("a{sv}"));
+  prepare_notification (&builder, notification, &fd_list);
+  g_variant_builder_close (&builder);
+
   if (callback)
     {
-      call_done_cb = call_done; 
+      call_done_cb = call_done;
       call_done_data = g_new (CallDoneData, 1);
       call_done_data->portal = g_object_ref (portal);
       call_done_data->callback = callback,
       call_done_data->data = data;
     }
 
-  g_dbus_connection_call (portal->bus,
-                          PORTAL_BUS_NAME,
-                          PORTAL_OBJECT_PATH,
-                          "org.freedesktop.portal.Notification",
-                          "AddNotification",
-                          g_variant_new ("(s@a{sv})", id, notification),
-                          NULL,
-                          G_DBUS_CALL_FLAGS_NONE,
-                          -1,
-                          cancellable,
-                          call_done_cb,
-                          call_done_data);
+  g_dbus_connection_call_with_unix_fd_list (portal->bus,
+                                            PORTAL_BUS_NAME,
+                                            PORTAL_OBJECT_PATH,
+                                            "org.freedesktop.portal.Notification",
+                                            "AddNotification",
+                                            g_variant_ref_sink (g_variant_builder_end (&builder)),
```

Comment:
The docs say that the `parameters` argument is owned by the caller, so this call won't unref the GVariant and it is leaked.

The docs also say "If the parameters GVariant is floating, it is consumed" and the `GVariantBuilder` docs say that `end` returns "A new, floating, GVariant".

So the best thing to do is this:

```suggestion
                                            g_variant_builder_end (&builder),
```